[apm-data] Apply lazy rollover on index template creation

axw opened the pull request:
Trigger a lazy rollover of existing data streams regardless of whether the index template is being created or updated. This ensures that the apm-data plugin will roll over data streams that were previously using the Fleet integration package.

Fixes https://github.com/elastic/elasticsearch/issues/116230

Commit:
Apply lazy rollover on index template creation

We should trigger a lazy rollover of existing data streams
regardless of whether the index template is being created
or updated. This ensures that the apm-data plugin will roll
over data streams that were previously using the Fleet
integration package.

Commit:
Add YAML REST test for template reinstallation

axw commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -793,41 +797,43 @@ private void invokeRollover(
-                            creationCheck.set(false);
-                            onRolloversBulkResponse(rolloverResponses);
-                        }
+            if (rolloverTargets.isEmpty()) {
+                creationCheck.set(false);
+                return;
+            }
```
This fixes another bug: we were never setting creationCheck to false in the case where there were no data streams to roll over on upgrade. This led to the plugin not recreating templates as intended, as the registry thought template creation was perpetually in-progress.

axw replied:
BTW without this change, the existing YAML REST tests would fail with the change to trigger the rollover path on template creation. I've also added a test case for template reinstallation for good measure.

gmarouli replied:
Nice catch! I would recommend to change the code a bit to make this less error prone. I think it will be easier if the responsibility to reset the check flag would happen in one spot. I think it would be nice to move to https://github.com/elastic/elasticsearch/pull/116219/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R552-R555.

For example:

```
 if (applyRolloverAfterTemplateV2Update()) {
       invokeRollover(state, templateName, indexTemplate, () -> creationCheck.set(false));
} else {
     creationCheck.set(false);
}
```

Then the `invokeRollover` would apply it accordingly:
```
    private void invokeRollover(
        final ClusterState state,
        final String templateName,
        final ComposableIndexTemplate indexTemplate,
        Runnable runAfter
    ) {
        final Executor executor = threadPool.generic();
        executor.execute(() -> {
            List<String> rolloverTargets = findRolloverTargetDataStreams(state, templateName, indexTemplate);
            if (rolloverTargets.isEmpty()) {
                runAfter.run();
                return;
            }
            ActionListener<RolloverResponse> groupedActionListener = ActionListener.runAfter(new GroupedActionListener<>(
                rolloverTargets.size(),
                new ActionListener<>() {
                    @Override
                    public void onResponse(Collection<RolloverResponse> rolloverResponses) {
                        onRolloversBulkResponse(rolloverResponses);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        onRolloverFailure(e);
                    }
                }), runAfter);
            for (String rolloverTarget : rolloverTargets) {
                logger.info(
                    "rolling over data stream [{}] lazily as a followup to the upgrade of the [{}] index template [{}]",
                    rolloverTarget,
                    getOrigin(),
                    templateName
                );
                RolloverRequest request = new RolloverRequest(rolloverTarget, null);
                request.lazy(true);
                request.masterNodeTimeout(TimeValue.MAX_VALUE);
                executeAsyncWithOrigin(
                    client.threadPool().getThreadContext(),
                    getOrigin(),
                    request,
                    groupedActionListener,
                    (req, listener) -> client.execute(RolloverAction.INSTANCE, req, listener)
                );
            }
        });
    }
```

The benefit of the above, is that `creationCheck` would be reset once when all the listeners have finished (assuming my understanding is correct) and the responsibility lies on the first snippet to declare that there is something that needs to be run and the `invokeRollover` needs to apply that `runAfter`, whatever that is.

Does this make sense? By the way this is a draft, if you like the suggestion please test it also ;)

axw replied:
:+1: I like it, thanks for the suggestion. I agree that it's a bit error prone (evidently!) with the way it's spread around at the moment. I will test and push an update shortly.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @axw, I've created a changelog YAML for you.

axw committed:
Update docs/changelog/116219.yaml

felixbarny reviewed (approved):
LGTM, but the data management team should also have a look.

elasticsearchmachine commented:
Hi @axw, I've updated the changelog YAML for you.

axw committed:
Update docs/changelog/116219.yaml

gmarouli commented on x-pack/plugin/apm-data/src/yamlRestTest/resources/rest-api-spec/test/10_rollover.yml:
```diff
@@ -0,0 +1,54 @@
+---
+setup:
+  - do:
+      cluster.health:
+        wait_for_events: languid
```
It is not clear to me why we are waiting here

axw replied:
The intention here is to wait for the apm-data plugin to install the templates. Is that not necessary, or is there a better way?

gmarouli replied:
In general this increases the chances that the templates would be installed before the test runs, but it doesn't guarantee it. So, in this specific case, I do not think it adds anything for two reasons:

1. The only test in this suite disables the apm plugin.
2. You skip the test anyway because we cannot guarantee that the templates will be installed, so why wait?

axw replied:
Fair. The real reason it's there is because I copied another test ;)
I'll remove the setup from this file.

gmarouli reviewed (commented):
Nice work @axw , in general it looks really good. I added some suggestions, let me know what you think.

Commit:
Code review suggestion

https://github.com/elastic/elasticsearch/pull/116219#discussion_r1828992554

gmarouli reviewed (approved):
LGTM, I would still recommend to remove the waiting in the `10_rollover.yml` test, but it's minor. Good work @axw 🚀

Commit:
Remove wait_for_events from setup

This doesn't guarantee the templates are set up,
it only increases the chances; and we disable the
plugin at the start of the test anyway.

lahsivjar reviewed (approved):
LGTM! one minor doubt/concern.

axw commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -550,8 +549,8 @@ private void putComposableTemplate(
                         if (response.isAcknowledged()) {
-                            if (isUpgrade && applyRolloverAfterTemplateV2Upgrade()) {
-                                invokeRollover(state, templateName, indexTemplate, creationCheck);
+                            if (applyRolloverAfterTemplateV2Update()) {
```
I'm not sure I understand the scenario. Do you mean a negative check, verifying something that does not match the index pattern is rolled over? I think the existing `IndexTemplateRegistryTests.testAutomaticRollover` covers that: it creates two logs and one traces data stream, and verifies that only the logs ones are rolled over when updating the logs template.

lahsivjar replied:
Not exactly a negative check -- my concern is when plugins operate under normal circumstances then there is no need for migrating datastream to the plugin. I was thinking of having a test case to cover this scenario. Does it make sense?

lahsivjar commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.15|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/116244">](https://github.com/elastic/elasticsearch/pull/116244)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.1 -->
